### PR TITLE
Fix various data issues

### PR DESCRIPTION
**Pull request by thibautjombart**
## Outline of the PR

Fixes various data-related issues which cause `asmodee` to crash or misbehave. This includes:

* missing data in the predictors
* missing data in the response variable
* categorical predictors (`factors`) with unknown levels in the testing set

The function `sanitize_models` performs the checks against the above, returning `TRUE` for models passing the checks. Only models passing the checks are now used in `asmodee`.

This also implements a `force_positive` argument, which ensures that model predictions are >= 0. This permits to use Gaussian models more safely - they turn out to be better options than Poisson GLMs in many cases.

## Things left to do
These will need to be treated as separate issues / PRs:

* rationalise `detect_changepoint`, which should really use `asmodee(..., fixed_k = xxx)`
* add tests for sanitation functions
* enable paralelisation for asmodee

### CODE REVIEW

File: R/get_predictor_data.R [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#' @param df A `data.frame` containing data required by the models (and possibly
+#'   non-relevant columns as well).
+#'
+#' @return A `data.frame` containing exactly the predictors of the model; if
+#'   some data are missing, an error is issued.
+#'
+get_predictor_data <- function(model, df) {
+  ## Extract variable names from the formula
+  form <- trending::get_formula(model)
+  vars <- unlist(strsplit(as.character(form)[3], "[-:+*|]"))
```
Comment (tjtnew) [inline R/get_predictor_data.R:21]: would something like `all.vars(form)[-1]` be better here?

File: R/sanitize_models.R [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                           x_training,
+                           x_testing,
+                           warn = FALSE) {
+    
+  # Auxiliary functions
+  
+  ## Check level consistency for a single factor
+  ##
+  ## Returns TRUE if levels are okay, FALSE otherwise
+  check_factor <- function(f_train, f_test) {
```
Comment (tjtnew) [inline R/sanitize_models.R:36]: Is there a reason to define this within sanitize_model?  This means it gets defined each time the function is called.

File: R/sanitize_models.R [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    all(levels(f_test) %in% levels(f_train))
+  }
+
+  ## Check level consistency for multiple factors
+  ##
+  ## This simply loops over all columns of testing data.frame, ensuring matching
+  ## names, and assuming all variables are factors.
+  check_factors <- function(df_train, df_test) {
```
Comment (tjtnew) [inline R/sanitize_models.R:44]: same as previous comment.

**Review by tjtnew — COMMENTED** [review]
A few minor comments left - nothing that stops it being merged.

File: R/get_predictor_data.R (hunk 1 above)
Comment (thibautjombart) [reply R/get_predictor_data.R:21]: Nice! Will push the change

File: R/sanitize_models.R (hunk 2 above)
Comment (thibautjombart) [reply R/sanitize_models.R:36]: Will move to separate files

File: R/sanitize_models.R (hunk 3 above)
Comment (thibautjombart) [reply R/sanitize_models.R:44]: same reply - push coming soon.

**Comment by thibautjombart** [comment]
Thanks for the review. Should be all fixed now.

**Pull request merged**